src/verify.c

```diff
@@ -412,6 +412,35 @@ static int get_required_files(struct manifest *official_manifest, struct list *s
 	return ret;
 }
 
+static void check_warn_freespace(const struct file *file)
+{
+	long fs_free;
+	char *original = NULL;
+	static bool no_freespace_flag = false;
+	struct stat st;
+
+	if (cmdline_option_install && no_freespace_flag == false) {
+		string_or_die(&original, "%s/staged/%s", state_dir, file->hash);
+		fs_free = get_available_space(path_prefix);
+		if (fs_free < 0 || stat(original, &st) != 0) {
+			fprintf(stderr, "\tWarning: Unable to determine free space on filesystem.\n");
+			goto out;
+		}
+		if (fs_free < st.st_size * 1.1) {
+			fprintf(stderr, "\tWarning: File to install (%s) too large by %ldK.\n",
+				file->filename, (st.st_size - fs_free) / 1000);
+			/* set flag to skip checking space on the second failure, assume we're still out of space */
+			no_freespace_flag = true;
+		}
+	} else {
+		fprintf(stderr, "\tWarning: No space left on device\n");
```

Comment:
So if `cmdline_option_install` is false this warning will be printed regardless of the `no_freespace_flag`?

Reply:
True - if you run 'verify --fix', and a file fails the final hash verification, this function will print a warning for no space; even if there is enough free space.

I've verified this happens and will fix. Thanks!

Reply:
What do you think about something like this:
`if (cmdline_option_install || cmdline_option_fix && no_freespace_flag == false)`
and then change the final printf to something like:
`fprintf(stderr, "\tContinuing with install or fix...\n");`

I think those are the only two paths that call add_missing_files().

It's also possible that stat() could fail - and the reported message isn't quite accurate (in that we could determine free space, but not how large the file we want to copy is). Should I change the output or add another if() to check stat separately?

Reply:
well `if (cmdline_option_install || cmdline_option_fix && no_freespace_flag == false)` will always pass if `cmdline_option_install` is true regardless of `no_freespace_flag`.

Why don't you convert this logic to a guard clause instead? Something like:
```C
static void thing(const struct file *file)
{
    if (!(cmdline_option_install || cmdline_option_fix)) {
        goto out;
    }

    if (no_freespace_flag) {
        fprintf(stderr, "Warning: ...");
        goto out;
    }

    /* the rest of your code */

out: 
    fprintf(stderr, "\tContinuing operation...\n");
    free_string(&original);
}

Reply:
This is a great idea for this case - the thought had honestly not crossed my mind... I think this is part of "knocking the rust off" :wink: